### PR TITLE
fix secret recovery and verification of signature using go-ethereum

### PR DESCRIPTION
**Pull request by noot**

### CODE REVIEW

File: secp256k1/dleq.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+// w := witness
+// prove R = w*G and R' = w*Y
+//
+// k := random scalar
+// Q := k*G
+// Q' := k*Y
+// z := hash( R || R' || Q || Q')
+// s := k + w*z
+// proof := (R, R', z, s)
+func dleqProve(w *secp256k1.Fn, R_a, R, Y *secp256k1.Point) (*dleqProof, error) {
```
Comment (jimjbrettj) [inline secp256k1/dleq.go:28]: I like that you moved these funcs, nice

File: secp256k1/keys.go [hunk 2]
```diff
@@ -64,6 +65,11 @@ type PublicKey struct {
 func (k *PublicKey) Encode() ([]byte, error) {
 	var b [33]byte
 	k.key.PutBytes(b[:])
+
+	// go-ethereum requires an encoded public key to specify its "type"
```
Comment (jimjbrettj) [inline secp256k1/keys.go:69]: ahhhhh

**Review by jimjbrettj — APPROVED** [review]
looks great! One question: where did we get the variable names from, are they based on something? Should we use same naming as the spec?

**Comment by noot** [comment]
@jimjbrettj which variable names exactly? a lot of them are just based on what they are usually called in crypto code, but maybe some can be improved?

**Pull request merged**